### ISSUE TITLE
Poor Performance on Filtered MassIVE-KB Data – Request for Guidance on Configuration Tuning

### ISSUE THREAD

**Issue by dumbgoos** [PHOENIXcenter/pi-PrimeNovo#9]
Hi,

I am writing to request your assistance regarding an issue I encountered while training Primenovo on a filtered MassIVE-KB dataset. I have carefully followed the configuration details provided, but unfortunately, the results are far from satisfactory.

**Dataset and Training Setup:**  
- **Dataset:** Filtered MassIVE-KB  
  - **Training spectra:** 1,818,143  
  - **Validation spectra:** 95,722  
  -  **data link**: https://drive.google.com/drive/folders/1Uurw4wiMa5_nhEPcjSAe9rNnTTo0LzLW?usp=drive_link
- **Hardware:** 8 A100 GPUs (40GB each)  
- **Training Parameters:**  
  - Batch size: 1600  
  - Learning rate: 0.0004  
- **Configuration:** I have attached the full configuration file for your reference.

**Observed Issues:**  
I have monitored the training performance using the following metrics, and both indicate significant issues:
- **test_pep_recall_epoch:** The recall performance appears to be much lower than expected.

![Image](https://github.com/user-attachments/assets/dca525d2-96da-4d92-a4b8-2191a264cd39)

- **test_aa_precision_epoch:** The amino acid precision is also underperforming considerably.

![Image](https://github.com/user-attachments/assets/eef18427-f28c-49dd-97ed-af1c4b05f6a0)

**Request for Assistance:**  
Given these challenges, I kindly ask for your guidance on how to adjust the configuration parameters to achieve better results. Specifically:
- Are there any recommended modifications to the learning rate, batch size, or other hyperparameters for this dataset?
- Could there be any issues related to the scaling of the training data or GPU utilization that might be affecting convergence?
- Any additional insights or best practices you could share for optimizing performance on such a large dataset would be greatly appreciated.

```yaml
###
# Casanovo configuration.
# Blank entries are interpreted as "None"
###

#idx_ratio = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.8, 1, 1]

# Random seed to ensure reproducible results.
random_seed: -1

#PMC decoding parameters---
PMC_enable: True
mass_control_tol: 0.1  #unit: Da, eveything decoded will be within 0.1 Da of MS/MS given mass

#--------
load_file_name: ""
# random_seed: -1 # Random
custom_ctc: False
# Spectrum processing options.
n_peaks: 150
min_mz: 1
max_mz: 6500.0
min_intensity: 0.0
remove_precursor_tol: 1  # Da
max_charge: 10
precursor_mass_tol: 10  # 50 ppm
isotope_error_range: [0, 1]

# Model architecture options.
dim_model:  400 #512
n_head: 8 #8
dim_feedforward: 1024
n_layers: 9
dropout: 0.18
dim_intensity:
custom_encoder:
max_length: 100 #70 #NON73: 32
residues:
  "G": 57.021464
  "A": 71.037114
  "S": 87.032028
  "P": 97.052764
  "V": 99.068414
  "T": 101.047670
  "C+57.021": 160.030649
  "L": 113.084064
  "I": 113.084064
  "N": 114.042927
  "D": 115.026943
  "Q": 128.058578
  "K": 128.094963
  "E": 129.042593
  "M": 131.040485
  "H": 137.058912
  "F": 147.068414
  "R": 156.101111
  "Y": 163.063329
  "W": 186.079313
  # Amino acid modifications.
  "M+15.995": 147.035400
  "N+0.984": 115.026943  
  "Q+0.984": 129.042594
  # N-terminal modifications.
  "+42.011": 42.010565
  "+43.006": 43.005814
  "-17.027": -17.026549
  "+43.006-17.027": 25.980265

n_log: 1
tb_summarywriter:

# Neptune logger
enable_neptune: True
neptune_project: 
neptune_api_token: 
tags: []

# Use epochs instead of iters
warmup_iters: 
max_iters: 

max_epochs: 150
warm_up_epochs: 5
learning_rate: 0.0004 
weight_decay: 8e-5 #8e-5
gradient_clip_val: 2.5
gradient_clip_algorithm: "norm"
accumulate_grad_batches: 1
sync_batchnorm: True
SWA: False # StochasticWeightAveraging
val_interval: 1
n_nodes: 1

# Training/inference options.
train_batch_size: 1600
predict_batch_size: 64
# n_beams: 5
n_beams: 5 # No beam search

logger:

num_sanity_val_steps: 0

train_from_scratch: True

save_model: True
train_from_resume: False
model_save_folder_path: "/home/dataset-fs-0/primenovo/model"
save_weights_only: False
every_n_train_steps: 50000
temp_dir_auto: True
```

Thank you very much for your time and assistance. I look forward to your suggestions and any guidance you can provide to help improve the training performance.

Best regards,  
Ling Luo
Xiamen University

**Comment by wensm77** [comment]
hi~
Does the data link you provide include all training data?

**Comment by dumbgoos** [comment]
> hi~ Does the data link you provide include all training data?

Yes!

**Comment by wyattz23** [comment]
Hi, training NAT requires a significant amount of parameter tunning and seed changing. One way to make your trianing more successful is to reduce the model size (dim, layer number  etc) as we notice larger models can be harder to converge (possible memorization issues in the early stage). 

We will release a stable training NAT model we just invented in near future, which is currently in submission, the new method guarantees training success, but before that i think you can try to reduce both model size and learning rate to achieve successful training.

**Comment by wyattz23** [comment]
and this is very common problem in our development as well.. not the issue with your data/GPUs. Try play with your parameter around and sometimes it will work..... Stay tune to our new NAT model with guaranteed training success...

**Comment by wensm77** [comment]
> > hi~ Does the data link you provide include all training data?
> 
> Yes!

Can you tell me how long it takes you to train this model on 8×A100?
thank!

**Comment by dumbgoos** [comment]
> > > hi~ Does the data link you provide include all training data?
> > 
> > 
> > Yes!
> 
> Can you tell me how long it takes you to train this model on 8×A100? thank!

Hi,

I sincerely apologize for the delay in my response. Thank you for your patience and for your interest in our work. We trained our model on 8×A100 GPUs for 50 epochs, and the entire process took nearly two days. Unfortunately, due to some issues with the parameter settings, the results were not as satisfactory as we had hoped, so we decided not to continue further training.

May I ask if you are also working on training primenovo? If so, I would be delighted to explore the possibility of collaborating and sharing our models.

Thank you very much for your understanding, and I look forward to your response.

Best regards
